mev-build-rs/src/reth_builder/builder.rs

```diff
@@ -124,19 +128,48 @@ impl<Pool, Client> Builder<Pool, Client> {
     }
 
     async fn on_epoch(&self, epoch: Epoch) {
-        // TODO: concurrent fetch
-        // TODO: batch updates to auction schedule
-        for relay in self.relays.iter() {
-            match relay.get_proposal_schedule().await {
-                Ok(schedule) => {
-                    let slots = self.auction_schedule.process(relay.clone(), &schedule);
-                    tracing::info!(epoch, ?slots, %relay, "processed proposer schedule");
-                }
-                Err(err) => {
-                    tracing::warn!(err = %err, "error fetching proposer schedule from relay")
-                }
-            }
+        // Multiple threads query relays and forward the schedule results
+        // to the receiver stream. The channel is bounded to 16, and will
+        // apply backpressure when needed.
+        let (tx, rx) = mpsc::channel::<(usize, Vec<ProposerSchedule>)>(16);
+        let mut rx = ReceiverStream::new(rx);
+
+        // Spawn tasks to fetch ProposerSchedule lists.
+        let tasks = self
+            .relays
+            .iter()
+            .enumerate()
+            .map(|(i, relay)| {
+                let relay = relay.clone();
+                let tx = tx.clone();
+                tokio::spawn(async move {
+                    let schedule = relay.get_proposal_schedule();
+                    let schedule = match schedule.await {
+                        Ok(schedule) => schedule,
+                        Err(e) => {
+                            tracing::error!(%e, "error fetching proposal schedule");
+                            return;
+                        }
+                    };
+                    // Send the schedules to the processing task.
+                    if let Err(e) = tx.send((i, schedule)).await {
+                        tracing::error!(%e, "error sending proposal schedule for processing");
+                    }
+                })
+            })
+            .collect::<FuturesUnordered<_>>();
+
+        // Stream the schedules for processing via the auction schedule.
+        while let Some((i, schedule)) = rx.next().await {
```

Comment:
`tx` needs to be dropped after creating `tasks` and before processing the `rx` stream, otherwise it will never close since not all senders will be dropped.